dask/dataframe/shuffle.py

```diff
@@ -963,57 +963,99 @@ def get_overlap(df, index):
         # we may lose rows (https://github.com/dask/dask/issues/6972).
-        if i == ddf.npartitions - 2 or ddf.divisions[i] != ddf.divisions[i + 1]:
-            frames.append((ddf._name, i))
+        if i == len(mins) - 2 or divisions[i] != divisions[i + 1]:
+            frames.append(dsk[(name, i)])
             dsk[(name, i)] = (methods.concat, frames)
             frames = []
 
     graph = HighLevelGraph.from_collections(name, dsk, dependencies=[ddf])
-    return new_dd_object(graph, name, ddf._meta, ddf.divisions)
+    return new_dd_object(graph, name, ddf._meta, divisions)
 
 
-def compute_and_set_divisions(df, **kwargs):
-    mins = df.index.map_partitions(M.min, meta=df.index)
-    maxes = df.index.map_partitions(M.max, meta=df.index)
-    mins, maxes = compute(mins, maxes, **kwargs)
+def _compute_partition_stats(column, allow_overlap=False, **kwargs) -> tuple:
+    """For a given column, compute the min, max, and len of each partition.
+
+    And make sure that the partitions are sorted relative to each other.
+    NOTE: this does not guarantee that every partition is internally sorted.
+    """
+    mins = column.map_partitions(M.min, meta=column)
+    maxes = column.map_partitions(M.max, meta=column)
+    lens = column.map_partitions(len, meta=column)
+    mins, maxes, lens = compute(mins, maxes, lens, **kwargs)
     mins = remove_nans(mins)
     maxes = remove_nans(maxes)
-
+    non_empty_mins = [m for m, l in zip(mins, lens) if l != 0]
+    non_empty_maxes = [m for m, l in zip(maxes, lens) if l != 0]
     if (
-        sorted(mins) != list(mins)
-        or sorted(maxes) != list(maxes)
-        or any(a > b for a, b in zip(mins, maxes))
+        sorted(non_empty_mins) != non_empty_mins
+        or sorted(non_empty_maxes) != non_empty_maxes
     ):
         raise ValueError(
-            "Partitions must be sorted ascending with the index", mins, maxes
+            f"Partitions must be sorted ascending by {column.name}",
+            f"In your dataset the mins of the partitions are: {mins}"
+            f"and the maxes are: {maxes}",
         )
+    if not allow_overlap and any(
+        a <= b for a, b in zip(non_empty_mins[1:], non_empty_maxes[:-1])
+    ):
+        raise ValueError(
+            "Partitions have overlapping values, so divisions are non-unique."
+            "Use `set_index(sorted=True)` with no `divisions` to allow dask to fix the overlap."
+        )
+    if not allow_overlap:
+        return (mins, maxes, lens)
+    else:
+        return (non_empty_mins, non_empty_maxes, lens)
 
-    df.divisions = tuple(mins) + (list(maxes)[-1],)
 
-    overlap = [i for i in range(1, len(mins)) if mins[i] >= maxes[i - 1]]
-    return fix_overlap(df, overlap) if overlap else df
+def compute_divisions(df, col=None, **kwargs) -> tuple:
+    column = df.index if col is None else df[col]
+    mins, maxes, _ = _compute_partition_stats(column, allow_overlap=False, **kwargs)
```

Comment:
Just wondering—maybe we'd want a `return_lengths=False` kwarg, so you can get the length of each partition out as well if you want it? In some rare cases, I could see that being valuable, and something you'd have to otherwise compute as a separate pass. For those cases, it might be nice to not just throw this information away.

Reply:
yeah maybe. I kind of like the `compute_current_divisions` returning just divisions. But honestly once I really thought about overlap and empty partitions, I am less confident that there are that many use cases where `compute_current_divisions` is all that helpful.